README.md
# Matrix Encrypt Attachment

Encrypt and decrypt attachments in accordance with the [Matrix](https://matrix.org) [Client-Server API spec](https://spec.matrix.org/v1.1/client-server-api/#sending-encrypted-attachments).

Supports browsers using [`WebCryptoAPI`](https://www.w3.org/TR/WebCryptoAPI/) and Node.js via [`crypto`](https://nodejs.org/api/crypto.html).

This project builds on [`browser-encrypt-attachment`](https://github.com/matrix-org/browser-encrypt-attachment) to add support for non-browser environments.

## Protocol versions

Previous versions of the protocol are not currently documented in the spec, and are described here for reference:

| Protocol | Description |
| --- | --- |
| v0 | use all 128 bits of the counter |
| v1 | use only 64 bits of the counter |
| v2 (current) | use only 64 bits and also zero out the other half to maximise the space before it wraps |

## Encryption

The library will encrypt to the following protocol versions:

| Protocol | Browser | Node.js |
| --- | --- | --- |
| Encrypt | v2 | v2 |

## Decryption

The library supports decryption of the following protocol versions:

| Protocol | Browser | Node.js |
| --- | --- | --- |
| Decrypt v0 | ✅ | ❌ |
| Decrypt v1 | ✅ | ❌ |
| Decrypt v2 | ✅ | ✅ |


CHANGELOG.md
## [1.0.2](https://github.com/matrix-org/matrix-encrypt-attachment/compare/v1.0.1...v1.0.2) (2022-01-29)


### Bug Fixes

* specify correct Node.js compatibility (>=12) ([999f197](https://github.com/matrix-org/matrix-encrypt-attachment/commit/999f1972f9ba708a5d2fd6f516347860c2a90aef))

## [1.0.1](https://github.com/matrix-org/matrix-encrypt-attachment/compare/v1.0.0...v1.0.1) (2022-01-14)


### Bug Fixes

* enabled automatic semantic release from main branch ([6b7fde1](https://github.com/matrix-org/matrix-encrypt-attachment/commit/6b7fde1f8529b141882a395adacbffc1c4f02a93))

# 1.0.0 (2022-01-13)

### BREAKING CHANGES

* Published as `matrix-encrypt-attachment` instead of `browser-encrypt-attachment`

### Features

* support for Node.js >=12 as well as browsers
* TypeScript typings are now provided

# 0.x.x

The following changes are archived from [`browser-encrypt-attachment`](https://www.npmjs.com/package/browser-encrypt-attachment):

Changes in [0.3.0](https://github.com/matrix-org/browser-encrypt-attachment/releases/tag/v0.3.0) (2016-11-23)
===================================================================================================
[Full Changelog](https://github.com/matrix-org/browser-encrypt-attachment/compare/v0.2.0...v0.3.0)

Changes:

 * Set the 64 bit counter to 0 to avoid overflow to make Android compatibility easier. (PR #4)

Changes in [0.2.0](https://github.com/matrix-org/browser-encrypt-attachment/releases/tag/v0.2.0) (2016-11-22)
===================================================================================================
[Full Changelog](https://github.com/matrix-org/browser-encrypt-attachment/compare/v0.1.0...v0.2.0)

Breaking changes:

 * Use a 64 bit counter in AES-CTR to make iOS compatibility easier. (PR #3)

Changes in [0.1.0](https://github.com/matrix-org/browser-encrypt-attachment/releases/tag/v0.1.0) (2016-11-11)
===================================================================================================
[Full Changelog](https://github.com/matrix-org/browser-encrypt-attachment/compare/v0.0.0...v0.1.0)

Breaking changes:

 * Use AES-CTR rather than AES-GCM to make iOS compatibility easier. (PR #2)


test/nodeHelper.js
/*
Copyright 2021-2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// For testing in Node.js we load the CommonJS module into global namespace so that we reuse the existing tests

/* eslint-disable */
const path = require('path');
global.MatrixEncryptAttachment = require(path.join('..', require('../package.json').main));
/* eslint-enable */


test/base64.Spec.js
/*
Copyright 2021-2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// eslint-disable-next-line valid-jsdoc
/**
 * Test if two things are equal after JSON stringification.
 * The arguments must be numbers, strings or arrays otherwise the behaviour is
 * undefined.
 */
function assertEq(got, want) {
    const gotJSON = JSON.stringify(got);
    const wantJSON = JSON.stringify(want);
    if (wantJSON != gotJSON) {
        throw new Error(`Want ${wantJSON} got ${gotJSON}`);
    }
}

describe('Base64', function() {
    const testVectors = [
        [[], ''],
        [[255], '/w'],
        [[255, 255], '//8'],
        [[255, 255, 255], '////'],
        [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 'AAECAwQFBgcICQ'],
    ];

    testVectors.forEach(function(vector) {
        const input = vector[0];
        const want = vector[1];
        it(`encodes ${JSON.stringify(input)} correctly`, function() {
            const got = MatrixEncryptAttachment.encodeBase64(new Uint8Array(input));
            assertEq(got, want);
        });
    });
    testVectors.forEach(function(vector) {
        const input = vector[1];
        const want = vector[0];
        it(`decodes ${JSON.stringify(input)} correctly`, function() {
            const got = Array.prototype.slice.call(MatrixEncryptAttachment.decodeBase64(input));
            assertEq(got, want);
        });
    });
});


test/decrypt.Spec.js
/*
Copyright 2021-2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

function assertEq(got, want) {
    const gotJSON = JSON.stringify(got);
    const wantJSON = JSON.stringify(want);
    if (wantJSON != gotJSON) {
        throw new Error(`Want ${wantJSON} got ${gotJSON}`);
    }
}

const v2TestVectors = [
    [
        '',
        {
            'iv': 'AAAAAAAAAAAAAAAAAAAAAA',
            'key': {
                'kty': 'oct',
                'key_ops': [
                    'encrypt',
                    'decrypt',
                ],
                'k': 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
                'alg': 'A256CTR',
            },
            'v': 'v2',
            'hashes': {
                'sha256': '47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU',
            },
        },
        '',
    ],
    [
        '5xJZTt5cQicm+9f4',
        {
            'iv': '//////////8AAAAAAAAAAA',
            'key': {
                'kty': 'oct',
                'key_ops': [
                    'encrypt',
                    'decrypt',
                ],
                'k': '__________________________________________8',
                'alg': 'A256CTR',
            },
            'v': 'v2',
            'hashes': {
                'sha256': 'YzF08lARDdOCzJpzuSwsjTNlQc4pHxpdHcXiD/wpK6k',
            },
        },
        'SGVsbG8sIFdvcmxk',
    ],
    [ // plain text YWxwaGF with protocol v2
        'zhtFStAeFx0s+9L/sSQO+WQMtldqYEHqTxMduJrCIpnkyer09kxJJuA4K+adQE4w+7jZe/vR9kIcqj9rOhDR8Q',
        {
            'iv': '//////////8AAAAAAAAAAA',
            'key': {
                'kty': 'oct',
                'key_ops': [
                    'encrypt',
                    'decrypt',
                ],
                'k': '__________________________________________8',
                'alg': 'A256CTR',
            },
            'v': 'v2',
            'hashes': {
                'sha256': 'IOq7/dHHB+mfHfxlRY5XMeCWEwTPmlf4cJcgrkf6fVU',
            },
        },
        'YWxwaGFudW1lcmljYWxseWFscGhhbnVtZXJpY2FsbHlhbHBoYW51bWVyaWNhbGx5YWxwaGFudW1lcmljYWxseQ',
    ],
];

const v1TestVectors = [
    [ // plain text YWxwaGF with protocol v1
        'tJVNBVJ/vl36UQt4Y5e5m84bRUrQHhcdLPvS/7EkDvlkDLZXamBB6k8THbiawiKZ5Mnq9PZMSSbgOCvmnUBOMA',
        {
            'iv': '/////////////////////w',
            'key': {
                'kty': 'oct',
                'key_ops': [
                    'encrypt',
                    'decrypt',
                ],
                'k': '__________________________________________8',
                'alg': 'A256CTR',
            },
            'v': 'v1',
            'hashes': {
                'sha256': 'LYG/orOViuFwovJpv2YMLSsmVKwLt7pY3f8SYM7KU5E',
            },
        },
        'YWxwaGFudW1lcmljYWxseWFscGhhbnVtZXJpY2FsbHlhbHBoYW51bWVyaWNhbGx5YWxwaGFudW1lcmljYWxseQ',
    ],
];

const v0TestVectors = [
    [ // plain text YWxwaGF with protocol v0
        'tJVNBVJ/vl36UQt4Y5e5myqUL3M8OtjRVQljZ+LlwbJeucRIM7CeKDJGGOjlJ1bqpqUdl6zytXJ3dCyvnUi4eQ',
        {
            'iv': '/////////////////////w',
            'key': {
                'kty': 'oct',
                'key_ops': [
                    'encrypt',
                    'decrypt',
                ],
                'k': '__________________________________________8',
                'alg': 'A256CTR',
            },
            'hashes': {
                'sha256': '/K4w3G4zlLK312k66KxNPKDkWCn2QAH5aphAkuncTrQ',
            },
        },
        'YWxwaGFudW1lcmljYWxseWFscGhhbnVtZXJpY2FsbHlhbHBoYW51bWVyaWNhbGx5YWxwaGFudW1lcmljYWxseQ',
    ],
];

const browserTestVectors = [
    ...v2TestVectors,
    ...v1TestVectors,
    ...v0TestVectors,
];

const nodejsTestVectors = [
    ...v2TestVectors,
];

describe('DecryptAttachment', function() {
    const decryptTestVectors = typeof window === 'undefined' ? nodejsTestVectors : browserTestVectors;
    decryptTestVectors.forEach(function(vector) {
        const inputCiphertext = vector[0];
        const inputInfo = vector[1];
        const want = vector[2];
        it(`decrypts ${inputInfo.v || 'v0'} ${JSON.stringify([inputCiphertext, inputInfo])}`, function() {
            return MatrixEncryptAttachment
                .decryptAttachment(MatrixEncryptAttachment.decodeBase64(inputCiphertext), inputInfo)
                .then(function(got) {
                    assertEq(MatrixEncryptAttachment.encodeBase64(new Uint8Array(got)), want);
                });
        });
    });
});


src/index.js
/**
 * Encrypt an attachment.
 * @param {ArrayBuffer} plaintextBuffer The attachment data buffer.
 * @return {Promise} A promise that resolves with an object when the attachment is encrypted.
 *      The object has a "data" key with an ArrayBuffer of encrypted data and an "info" key
 *      with an object containing the info needed to decrypt the data.
 */
function encryptAttachment(plaintextBuffer) {
    let cryptoKey; // The AES key object.
    let exportedKey; // The AES key exported as JWK.
    let ciphertextBuffer; // ArrayBuffer of encrypted data.
    let sha256Buffer; // ArrayBuffer of digest.
    // Generate an IV where the first 8 bytes are random and the high 8 bytes
    // are zero. We set the counter low bits to 0 since it makes it unlikely
    // that the 64 bit counter will overflow.
    const ivArray = new Uint8Array(16); // Uint8Array of AES IV
    window.crypto.getRandomValues(ivArray.subarray(0, 8));
    // Load the encryption key.
    return window.crypto.subtle.generateKey(
        { 'name': 'AES-CTR', 'length': 256 }, true, ['encrypt', 'decrypt'],
    ).then(function(generateKeyResult) {
        cryptoKey = generateKeyResult;
        // Export the Key as JWK.
        return window.crypto.subtle.exportKey('jwk', cryptoKey);
    }).then(function(exportKeyResult) {
        exportedKey = exportKeyResult;
        // Encrypt the input ArrayBuffer.
        // Use half of the iv as the counter by setting the "length" to 64.
        return window.crypto.subtle.encrypt(
            { name: 'AES-CTR', counter: ivArray, length: 64 }, cryptoKey, plaintextBuffer,
        );
    }).then(function(encryptResult) {
        ciphertextBuffer = encryptResult;
        // SHA-256 the encrypted data.
        return window.crypto.subtle.digest('SHA-256', ciphertextBuffer);
    }).then(function(digestResult) {
        sha256Buffer = digestResult;

        return {
            data: ciphertextBuffer,
            info: {
                v: 'v2',
                key: exportedKey,
                iv: encodeBase64(ivArray),
                hashes: {
                    sha256: encodeBase64(new Uint8Array(sha256Buffer)),
                },
            },
        };
    });
}

/**
 * Decrypt an attachment.
 * @param {ArrayBuffer} ciphertextBuffer The encrypted attachment data buffer.
 * @param {Object} info The information needed to decrypt the attachment.
 * @param {Object} info.key AES-CTR JWK key object.
 * @param {string} info.iv Base64 encoded 16 byte AES-CTR IV.
 * @param {string} info.hashes.sha256 Base64 encoded SHA-256 hash of the ciphertext.
 * @return {Promise} A promise that resolves with an ArrayBuffer when the attachment is decrypted.
 */
function decryptAttachment(ciphertextBuffer, info) {
    if (info === undefined || info.key === undefined || info.iv === undefined
        || info.hashes === undefined || info.hashes.sha256 === undefined) {
        throw new Error('Invalid info. Missing info.key, info.iv or info.hashes.sha256 key');
    }

    let cryptoKey; // The AES key object.
    const ivArray = decodeBase64(info.iv);
    const expectedSha256base64 = info.hashes.sha256;
    // Load the AES from the "key" key of the inf bao object.
    return window.crypto.subtle.importKey(
        'jwk', info.key, { 'name': 'AES-CTR' }, false, ['encrypt', 'decrypt'],
    ).then(function(importKeyResult) {
        cryptoKey = importKeyResult;
        // Check the sha256 hash
        return window.crypto.subtle.digest('SHA-256', ciphertextBuffer);
    }).then(function(digestResult) {
        if (encodeBase64(new Uint8Array(digestResult)) != expectedSha256base64) {
            throw new Error('Mismatched SHA-256 digest');
        }
        let counterLength;
        if (info.v == 'v1' || info.v == 'v2') {
            // Version 1 and 2 use a 64 bit counter.
            counterLength = 64;
        } else {
            // Version 0 uses a 128 bit counter.
            counterLength = 128;
        }
        return window.crypto.subtle.decrypt(
            { name: 'AES-CTR', counter: ivArray, length: counterLength }, cryptoKey, ciphertextBuffer,
        );
    });
}

/**
 * Encode a typed array of uint8 as base64.
 * @param {Uint8Array} uint8Array The data to encode.
 * @return {string} The base64 without padding.
 */
function encodeBase64(uint8Array) {
    // Misinterpt the Uint8Array as Latin-1.
    // window.btoa expects a unicode string with codepoints in the range 0-255.
    const latin1String = String.fromCharCode.apply(null, uint8Array);
    // Use the builtin base64 encoder.
    const paddedBase64 = window.btoa(latin1String);
    // Calculate the unpadded length.
    const inputLength = uint8Array.length;
    const outputLength = 4 * Math.floor((inputLength + 2) / 3) + (inputLength + 2) % 3 - 2;
    // Return the unpadded base64.
    return paddedBase64.slice(0, outputLength);
}

/**
 * Decode a base64 string to a typed array of uint8.
 * This will decode unpadded base64, but will also accept base64 with padding.
 * @param {string} base64 The unpadded base64 to decode.
 * @return {Uint8Array} The decoded data.
 */
function decodeBase64(base64) {
    // Pad the base64 up to the next multiple of 4.
    const paddedBase64 = base64 + '==='.slice(0, (4 - base64.length % 4) % 4);
    // Decode the base64 as a misinterpreted Latin-1 string.
    // window.atob returns a unicode string with codepoints in the range 0-255.
    const latin1String = window.atob(paddedBase64);
    // Encode the string as a Uint8Array as Latin-1.
    const uint8Array = new Uint8Array(latin1String.length);
    for (let i = 0; i < latin1String.length; i++) {
        uint8Array[i] = latin1String.charCodeAt(i);
    }
    return uint8Array;
}

try {
    // eslint-disable-next-line no-undef
    exports.encryptAttachment = encryptAttachment;
    // eslint-disable-next-line no-undef
    exports.decryptAttachment = decryptAttachment;
} catch (e) {
    // Ignore unknown variable "exports" errors when this is loaded directly into a browser
    // This means that we can test it without having to use browserify.
    // The intention is that the library is used using browserify.
}


src/index.ts
/*
Copyright 2021-2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as webcrypto from './webcrypto';
import * as nodejs from './nodejs';

const hasWebcrypto = !!(typeof window !== 'undefined' && window.crypto?.subtle);

/**
 * Represents an `EncryptedFile` as described by https://spec.matrix.org/v1.1/client-server-api/#extensions-to-mroommessage-msgtypes
 */
export interface IEncryptedFile {
    /**
     * A JSON Web Key object containing the key used.
     */
    key: IEncryptedFileJWK;

    /**
     * The 128-bit unique counter block used by AES-CTR, encoded as unpadded base64.
     */
    iv: string;

    /**
     * Version of the encrypted attachments protocol. See README.md for supported protocol versions.
     */
    v?: string;

    /**
     * A map from an algorithm name to a hash of the ciphertext, encoded as unpadded base64. Clients should support the SHA-256 hash, which uses the key sha256.
     */
    hashes?: {
        sha256?: string;
    };
}

/**
 * Representation of `JWK` as described by https://spec.matrix.org/v1.1/client-server-api/#extensions-to-mroommessage-msgtypes
 */
export interface IEncryptedFileJWK extends JsonWebKey {
    /**
     * Key type. Must be `oct`.
     */
    kty: string;

    /**
     *  Key operations. Must at least contain `encrypt` and `decrypt`.
     */
    key_ops: string[];

    /**
     * Algorithm. Must be `A256CTR`.
     */
    alg: string;

    /**
     *  The key, encoded as urlsafe unpadded base64.
     */
    k: string;

    /**
     * Extractable. Must be `true`. This is a W3C extension.
     */
    ext: boolean;
}

/**
 * Encrypt an attachment to latest protocol specification (currently v2).
 * @param {ArrayBuffer} plaintextBuffer The attachment data buffer.
 * @return {Promise} A promise that resolves with an object when the attachment is encrypted.
 *      The object has a "data" key with an ArrayBuffer of encrypted data and an "info" key
 *      with an object containing the info needed to decrypt the data.
 */
export async function encryptAttachment(plaintextBuffer: ArrayBuffer): Promise<{
    data: ArrayBuffer;
    info: IEncryptedFile;
}> {
    return hasWebcrypto ? webcrypto.encryptAttachment(plaintextBuffer)
        : nodejs.encryptAttachment(Buffer.from(plaintextBuffer));
}

/**
 * Decrypt an attachment that has been encrypted with v1 or v2.
 * @param {ArrayBuffer} ciphertextBuffer The encrypted attachment data buffer.
 * @param {IEncryptedFile} info The information needed to decrypt the attachment.
 * @return {Promise<ArrayBuffer>} A promise that resolves with an ArrayBuffer when the attachment is decrypted.
 */
export async function decryptAttachment(ciphertextBuffer: ArrayBuffer, info: IEncryptedFile): Promise<ArrayBuffer> {
    return hasWebcrypto ? webcrypto.decryptAttachment(ciphertextBuffer, info)
        : nodejs.decryptAttachment(Buffer.from(ciphertextBuffer), info);
}

/**
 * Encode a typed array of uint8 as unpadded base64.
 * @param {Uint8Array} uint8Array The data to encode.
 * @return {string} The base64 without padding.
 */
export function encodeBase64(uint8Array: Uint8Array): string {
    return hasWebcrypto ? webcrypto.encodeBase64(uint8Array)
        : nodejs.encodeBase64(uint8Array);
}

/**
 * Decode a base64 string to a typed array of uint8.
 * This will decode unpadded base64, but will also accept base64 with padding.
 * @param {string} base64 The unpadded base64 to decode.
 * @return {Uint8Array} The decoded data.
 */
export function decodeBase64(base64: string): Uint8Array {
    return hasWebcrypto ? webcrypto.decodeBase64(base64)
        : nodejs.decodeBase64(base64);
}


karma.conf.js
// Karma configuration
// Generated on Tue Nov 01 2016 11:48:39 GMT+0000 (GMT)

module.exports = function(config) {
    config.set({

        // base path that will be used to resolve all patterns (eg. files, exclude)
        basePath: '',

        // frameworks to use
        // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
        frameworks: ['mocha'],

        // list of files / patterns to load in the browser
        files: [
            require('./package.json').browser,
            'test/**/*Spec.js',
        ],

        // list of files to exclude
        exclude: [
        ],

        // preprocess matching files before serving them to the browser
        // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
        preprocessors: {
        },

        // test results reporter to use
        // possible values: 'dots', 'progress'
        // available reporters: https://npmjs.org/browse/keyword/karma-reporter
        reporters: ['progress'],

        // web server port
        port: 9876,

        // enable / disable colors in the output (reporters and logs)
        colors: true,

        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: config.LOG_INFO,

        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: false,

        // start these browsers
        // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
        browsers: ['ChromeHeadless', 'FirefoxHeadless'],

        // Continuous Integration mode
        // if true, Karma captures browsers, runs the tests and exits
        singleRun: true,

        // Concurrency level
        // how many browser should be started simultaneous
        concurrency: Infinity,
    });
};


src/webcrypto.ts
/*
Copyright 2021-2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { IEncryptedFile, IEncryptedFileJWK } from '.';

export async function encryptAttachment(plaintextBuffer: ArrayBuffer): Promise<{
    data: ArrayBuffer;
    info: IEncryptedFile;
}> {
    // Generate an IV where the first 8 bytes are random and the high 8 bytes
    // are zero. We set the counter low bits to 0 since it makes it unlikely
    // that the 64 bit counter will overflow.
    const ivArray = new Uint8Array(16); // Uint8Array of AES IV
    window.crypto.getRandomValues(ivArray.subarray(0, 8));
    // Load the encryption key.
    const cryptoKey = await window.crypto.subtle.generateKey(
        { 'name': 'AES-CTR', 'length': 256 }, true, ['encrypt', 'decrypt'],
    );
    // Export the Key as JWK.
    const exportedKey = await window.crypto.subtle.exportKey('jwk', cryptoKey);
    // Encrypt the input ArrayBuffer.
    // Use half of the iv as the counter by setting the "length" to 64.
    const ciphertextBuffer = await window.crypto.subtle.encrypt(
        { name: 'AES-CTR', counter: ivArray, length: 64 }, cryptoKey, plaintextBuffer,
    );
    // SHA-256 the encrypted data.
    const sha256Buffer = await window.crypto.subtle.digest('SHA-256', ciphertextBuffer);
    return {
        data: ciphertextBuffer,
        info: {
            v: 'v2',
            key: exportedKey as IEncryptedFileJWK,
            iv: encodeBase64(ivArray),
            hashes: {
                sha256: encodeBase64(new Uint8Array(sha256Buffer)),
            },
        },
    };
}

export async function decryptAttachment(ciphertextBuffer: ArrayBuffer, info: IEncryptedFile): Promise<ArrayBuffer> {
    if (info === undefined || info.key === undefined || info.iv === undefined
        || info.hashes === undefined || info.hashes.sha256 === undefined) {
        throw new Error('Invalid info. Missing info.key, info.iv or info.hashes.sha256 key');
    }

    if (info.v && !info.v.match(/^v[1-2]$/)) {
        throw new Error(`Unsupported protocol version: ${info.v}`);
    }

    const ivArray = decodeBase64(info.iv);
    const expectedSha256base64 = info.hashes.sha256;
    // Load the AES from the "key" key of the inf bao object.
    const cryptoKey = await window.crypto.subtle.importKey(
        'jwk', info.key, { 'name': 'AES-CTR' }, false, ['encrypt', 'decrypt'],
    );
    const digestResult = await window.crypto.subtle.digest('SHA-256', ciphertextBuffer);
    if (encodeBase64(new Uint8Array(digestResult)) != expectedSha256base64) {
        throw new Error('Mismatched SHA-256 digest');
    }
    let counterLength: number;
    if (info.v == 'v1' || info.v == 'v2') {
        // Version 1 and 2 use a 64 bit counter.
        counterLength = 64;
    } else {
        // Version 0 uses a 128 bit counter.
        counterLength = 128;
    }
    return window.crypto.subtle.decrypt(
        { name: 'AES-CTR', counter: ivArray, length: counterLength }, cryptoKey, ciphertextBuffer,
    );
}

export function encodeBase64(uint8Array: Uint8Array): string {
    // Misinterpt the Uint8Array as Latin-1.
    // window.btoa expects a unicode string with codepoints in the range 0-255.
    const latin1String = String.fromCharCode.apply(null, uint8Array);
    // Use the builtin base64 encoder.
    const paddedBase64 = window.btoa(latin1String);
    // Calculate the unpadded length.
    const inputLength = uint8Array.length;
    const outputLength = 4 * Math.floor((inputLength + 2) / 3) + (inputLength + 2) % 3 - 2;
    // Return the unpadded base64.
    return paddedBase64.slice(0, outputLength);
}

export function decodeBase64(base64: string): Uint8Array {
    // Pad the base64 up to the next multiple of 4.
    const paddedBase64 = base64 + '==='.slice(0, (4 - base64.length % 4) % 4);
    // Decode the base64 as a misinterpreted Latin-1 string.
    // window.atob returns a unicode string with codepoints in the range 0-255.
    const latin1String = window.atob(paddedBase64);
    // Encode the string as a Uint8Array as Latin-1.
    const uint8Array = new Uint8Array(latin1String.length);
    for (let i = 0; i < latin1String.length; i++) {
        uint8Array[i] = latin1String.charCodeAt(i);
    }
    return uint8Array;
}


src/nodejs.ts
/*
Copyright 2021-2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import crypto from 'crypto';
import { IEncryptedFile, IEncryptedFileJWK } from '.';

export async function encryptAttachment(plaintextBuffer: Buffer): Promise<{
    data: ArrayBuffer;
    info: IEncryptedFile;
}> {
    // Generate an IV where the first 8 bytes are random and the high 8 bytes
    // are zero. We set the counter low bits to 0 since it makes it unlikely
    // that the 64 bit counter will overflow.
    const ivArray = crypto.randomBytes(16); // Uint8Array of AES IV
    ivArray.fill(0, 8);

    // generate 256 bit key
    const cryptoKey = crypto.randomBytes(32);

    // Export the Key as JWK.
    const exportedKey: IEncryptedFileJWK = {
        kty: 'oct',
        key_ops: ['encrypt', 'decrypt'],
        alg: 'A256CTR',
        // node 14+ supports base64url encoding directly, but node 12 doesn't so we implement the mapping ourselves
        k: Buffer.from(cryptoKey).toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, ''), // URL safe base64 without padding
        ext: true,
    };

    // Encrypt the input Buffer.
    const cipher = crypto.createCipheriv('aes-256-ctr', cryptoKey, ivArray);
    const ciphertextBuffer = Buffer.concat([
        cipher.update(plaintextBuffer),
        cipher.final(),
    ]);

    // SHA-256 the encrypted data.
    const sha256Buffer = crypto.createHash('sha256').update(ciphertextBuffer).digest();

    return {
        data: ciphertextBuffer,
        info: {
            v: 'v2',
            key: exportedKey,
            iv: encodeBase64(ivArray),
            hashes: {
                sha256: encodeBase64(sha256Buffer),
            },
        },
    };
}

// this is based on https://github.com/matrix-org/matrix-content-scanner/blob/main/src/decrypt.js
export function decryptAttachment(dataBuffer: Buffer, info: IEncryptedFile): Buffer {
    if (info === undefined || info.key === undefined || info.iv === undefined ||
        info.hashes === undefined || info.hashes.sha256 === undefined) {
        throw new Error('Invalid info. Missing info.key, info.iv or info.hashes.sha256 key');
    }

    if (info.v !== 'v2') {
        throw new Error(`Unsupported protocol version: ${info.v ?? 'v0'}`);
    }

    const expectedSha256base64 = info.hashes.sha256;

    // Convert from JWK to openssl algorithm
    // See https://www.w3.org/2012/webcrypto/wiki/KeyWrap_Proposal#JSON_Web_Key
    const algorithms = {
        'oct': {
            'A256CTR': 'aes-256-ctr',
        },
    };

    const alg = algorithms[info.key.kty] ? algorithms[info.key.kty][info.key.alg] : undefined;

    if (!alg) {
        throw new Error(
            `Unsupported key type/algorithm: ` +
            `key.kty = ${info.key.kty}, kry.alg = ${info.key.alg}`);
    }

    const key = decodeBase64(info.key.k);

    // Calculate SHA 256 hash, encode as base64 without padding
    const hashDigestBase64 = encodeBase64(crypto.createHash('sha256').update(dataBuffer).digest());

    if (hashDigestBase64 !== expectedSha256base64) {
        throw new Error('Unexpected sha256 hash of encrypted data');
    }

    const iv = decodeBase64(info.iv);

    const decipher = crypto.createDecipheriv(alg, key, iv);

    return Buffer.concat([
        decipher.update(dataBuffer),
        decipher.final(),
    ]);
}

export function encodeBase64(uint8Array: Uint8Array): string {
    const padded = Buffer.from(uint8Array).toString('base64');

    // remove padding
    const inputLength = uint8Array.length;
    const outputLength = 4 * Math.floor((inputLength + 2) / 3) + (inputLength + 2) % 3 - 2;
    // Return the unpadded base64.
    return padded.slice(0, outputLength);
}

export function decodeBase64(base64: string): Uint8Array {
    // add padding if needed
    const paddedBase64 = base64 + '==='.slice(0, (4 - base64.length % 4) % 4);
    return Buffer.from(paddedBase64, 'base64');
}
